### ISSUE TITLE
GPG + Yubikey -- SSH

### ISSUE THREAD

**Issue by naggie** [naggie/dotfiles#120]
My use cases
* Portable SSH identity (with forwarding)
* Portable GPG identity (with forwarding)
* Use of https://www.passwordstore.org/ sync'd with git

Future use cases
* Signing things
* Encrypting things (including email)
* Decrypting things

-----

Comprehensive guide: https://github.com/drduh/YubiKey-Guide

quick guide https://hugotunius.se/2018/07/13/yubikey-ssh-authentication.html

Good overview of GPG keys and smartcards in general https://gist.github.com/ageis/14adc308087859e199912b4c79c4aaa4

- [x] gpg as agent
- [x] ~force console based askpass (if at all, and a non broken version) https://blog.separateconcerns.com/2014-10-24-cli-passwords.html https://unix.stackexchange.com/questions/371901/gpg-agent-refuses-ssh-keys-with-ssh-add-reporting-agent-refused-operation may need update tty pinentry is so very broken (wrong pane, only captures some, should use tmux root tty) could sigstop/cont tmux so pinentry-curses can work.~
- [x] pinentry with ssh is screwed -- wrong tty and mixed I/O -- maybe use gui or gpg forwarded agent instead of ssh agent protocol over the network. Another solution could be to unlock card before ssh (pin cached by yubikey for enture session) fix --  use GUI
- [x] timeout for pinentry (to avoid using an attempt if it's screwed up)
- [x] look into enabling touch to use  https://github.com/drduh/YubiKey-Guide#touch-to-authenticate
- [x] change admin pin codes
- [x] change user pin codes
- [x] decide on key topology
- [x] generate keys
- [x] generate SSH publkey, add to dotfiles add to install-naggie
- [x] generate GPG pubkey (of main key or all?) add to install-naggie
- [x] import both pubkeys to gpg (ultimate trust) with install-naggie
- [x] ~either do the same with stubs, or xfer from card somehow automatically~ stubs not necessary if pubkeys are ultimately trusted -- done via gpg..conf
- [x] update github with ssh/gpg pubkeys
- [x] also forward GPG agent... (needs reliable socket reaping) maybe with a mv to secondary location
- [x] ...enforce github signing of commits
- [x] configure a keyserver maybe http://keys.gnupg.net/ (there is already a sane default)
- [x] stop apple keychain and gnome keychain from getting involved https://wiki.gnupg.org/GnomeKeyring https://blog.josefsson.org/2015/01/02/openpgp-smartcards-and-gnome/
- [x] publish all pubkeys to keyserver
- [x] try on a new PC
- [x] don't start up remote gpg agent (also see what it does when agent forwarding has happened) --- in fact stop it
- [x] key generation script https://www.gnupg.org/documentation/manuals/gnupg/Unattended-GPG-key-generation.html
- [x] install.sh should restart gpg agent to enable new forwarding socket if it's not there (local only)
- [x] gpgagent is run by systemd user service on ubuntu. Kill with fire.
- [x] consider a pre-ssh command to remove the socket, kill gpg, find socket path. May help make work with non-configured server.
- [x] agent forwarding security mitigation -- touch to allow, explicit ssh command (gssh) or something
- [x] sort out issues with gpg2 vs gpg2 ubuntu 16.04
- [x] work around possible bug -- see section on sleepwalker https://evilmartians.com/chronicles/stick-with-security-yubikey-ssh-gnupg-macos (I did not experience said bug)
- [ ] support gssh chaining (command that calls func or something)
- [ ] install ykman on all platforms  https://github.com/Yubico/yubikey-manager
- [ ] wire up https://github.com/maximbaz/yubikey-touch-detector or equivalent to a notification and/or sound?
- [x] --card-status required to initallly discover card's secret keys (and bind card to stub)
- [x] set up guide (with recommendation for touch settings)

**Comment by naggie** [comment]
https://hugotunius.se/2018/07/13/yubikey-ssh-authentication.html

**Comment by naggie** [comment]
https://gist.github.com/ageis/14adc308087859e199912b4c79c4aaa4

**Comment by mafredri** [comment]
Saw this issue linked from the one about Pinentry in Pure, thought I'd let you know I'm working on something that might interest you, [gpg-notify](https://github.com/mafredri/gpg-notify), still very early days though. Mostly a POC stage atm. It works via parsing `gpg-agent` logs (and working on parsing `scdaemon` as well for more exact results).

When I started using gpg and YubiKeys I noticed I needed more information about what's going on, so it notifies about smartcard events. It also tries to look up what processes have initiated the communication with gpg-agent and gives information about them (although atm I think it just logs most of it to the terminal). Im also planning on showing what processes generally have initiated pinentry, etc.
